File: stock_search/data_sources/stockanalysis/page_scrapers/financials.py
"""StockAnalysis financials page scraping helpers."""

from __future__ import annotations

from collections.abc import Callable

from selectolax.lexbor import LexborHTMLParser

from ..parsing import build_model_from_rows, extract_table_rows
from ..schemas import StockAnalysisFinancials

STOCKANALYSIS_FINANCIALS_URL = "https://stockanalysis.com/stocks/{ticker}/financials/"

FINANCIALS_FIELD_SPECS = {
    "revenue_growth": ("Revenue Growth (YoY)", "parse_percent_ratio"),
    "eps_diluted": ("EPS (Diluted)", "parse_number"),
    "eps_growth": ("EPS Growth", "parse_percent_ratio"),
    "gross_margin": ("Gross Margin", "parse_percent_ratio"),
    "operating_margin": ("Operating Margin", "parse_percent_ratio"),
}


def scrape_financials_snapshot(
    *,
    ticker_lower: str,
    fetch_soup: Callable[[str], LexborHTMLParser | None],
) -> StockAnalysisFinancials:
    """Scrape the StockAnalysis financials table into a snapshot."""
    financials_url = STOCKANALYSIS_FINANCIALS_URL.format(ticker=ticker_lower)
    soup = fetch_soup(financials_url)
    if soup is None:
        return StockAnalysisFinancials()

    rows = extract_table_rows(soup, cell_selector="td", keep_first=False)
    return build_model_from_rows(
        rows,
        model_type=StockAnalysisFinancials,
        field_specs=FINANCIALS_FIELD_SPECS,
    )

File: src/stock-search/data-sources/stockanalysis/page-scrapers/financials.ts
/** StockAnalysis financials page scraping helpers. */

import type { StockAnalysisFinancialsPayload } from "../parsing.js";
import { STOCKANALYSIS_FINANCIALS_URL } from "../urls.js";
import {
	evaluateObjectLiteral,
	extractObjectLiteral,
	fetchText,
	toPercent,
} from "../../shared.js";

/** Scrape the StockAnalysis financials page into a structured snapshot. */
export async function scrapeFinancialsSnapshot(
	tickerLower: string,
): Promise<Record<string, unknown>> {
	const html = await fetchText(
		STOCKANALYSIS_FINANCIALS_URL.replace("{ticker}", tickerLower),
	);
	const payload = evaluateObjectLiteral<StockAnalysisFinancialsPayload>(
		extractObjectLiteral(html ?? "", "financialData"),
	);
	if (!payload) {
		return {};
	}

	return {
		revenue_growth: toPercent(payload.revenueGrowth?.[0] ?? null),
		eps_growth: toPercent(payload.epsGrowth?.[0] ?? null),
		gross_margin: toPercent(payload.grossMargin?.[0] ?? null),
	};
}
